feat: persistent user home documentation (#2822)
codenvy/che-docs: modules/administration-guide/pages/about-persistent-user-home.adoc

:_content-type: CONCEPT
:description: About persistent user home
:keywords: administration guide, about, {prod-id-short}, persistent, user, home

> :navtitle: About persistent user home

Reviewer: ```suggestion
:navtitle: Persistent user home
```

:page-aliases:

[id="about-persistent-user-home"]

> = About persistent user home

Reviewer: ```suggestion
= Persistent user home
```

{prod} provides a persistent home directory feature that allows each non-ephemeral workspace to have their `/home/user` directory be persisted across workspace restarts.

> This feature can be enabled in the CheCluster by setting `spec.devEnvironments.persistUserHome.enabled` to `true`.

Reviewer: ```suggestion
You can enable this feature in the CheCluster by setting `spec.devEnvironments.persistUserHome.enabled` to `true`.
```

For newly started workspaces, this feature creates a PVC mounted to the `/home/user` path of the tools container.
In this documentation, a "tools container" will be used to refer to the first container in the devfile.
This container is the container that includes the project source code by default.

> When the PVC is mounted for the first time, the persistent volume contents are empty and therefore must be populated with the `/home/user` folder content.

Reviewer 2: `persistent volume contents` -> `persistent volume's content` (needs an `'s`)

Author: Fixed

By default, the `persistUserHome` feature creates an init container for each new workspace pod named `init-persistent-home`.

> This init container is created with the tools container image and is responsible for running a `stow` command to create symbolic links

Reviewer 2: I don't know if it's worth mentioning that "For files that cannot be symbolically linked to the `/home/user/ `directory, `cp` is used". Though, this might be a low level [technical detail](https://github.com/devfile/devworkspace-operator/blob/7ecb9855b53b0cb6a22f1f17dd07fc0fafd0b426/pkg/library/home/persistentHome.go#L36-L38) that is not as relevant in the docs.

Author: I've added this in my latest commit

> in the persistent volume to populate the `/home/user` folder.

Reviewer 2: Nitpick: I think the standard in Unix (and Linux?) is to use "directory" instead of "folder". But readers will understand the meaning of both.

However... we're already inconsistently using a mix of "directory" and "folder" in the Che docs. So this is really not a big deal.

Author: Updated to directory

The primary function of the `stow` command is to run:
[subs="+quotes,attributes"]
----
stow -t /home/user/ -d /home/tooling/ --no-folding
----

> The command above creates symbolic links in `/home/user` for files and folders located in /home/tooling. This populates the persistent volume with symbolic links to the content in /home/tooling. As a result, it the `persistentUserHome` feature expects the tooling image to have it's `/home/user/` content within `/home/tooling`.

Reviewer 2: The  /home/tooling is missing the surrounding backticks ` twice in the first two sentences (for consistency).

Also `it's` -> `its`:
`...expects the tooling image to have its `/home/user/` content...`

Author: Fixed

For example, if the tools container image contains files in the `home/tooling` folder such as `.config` and `.config-folder/another-file`, stow will create symbolic links in the persistent volume in the following manner:

.Tools container with `persistUserHome` enabled
image::persistent-user-home/tools-container-example.png[Persistent user home example scenario]

> The init container writes the output of the `stow` command to `/home/user/.stow.log` and will only run `stow` the first time the PVC is mounted to the workspace. 

Reviewer 2: Technical nitpick: Isin't the stow command run only the first time when the workspace is started up? When aa workspace pod gets terminated (if the workspace shuts down) won't it unmount to the PVC and then get mounted again?

Author: Right, what I mean is it runs the first time it's mounted, in the second/third/fourth workspace starts, it's the second/third/fourth time it's mounted, so it doesn't run. 

Is the wording is a bit unclear?

In my latest commit, I've updated it to:
```
The init container writes the output of the `stow` command to `/home/user/.stow.log`
and will only run `stow` the first time the persistent volume is mounted to the workspace. 
```

> Using the `stow` command to populate `/home/user` content in the persistent volume provides two main advantages.

Reviewer 2: Maybe this sentence should end with a `:` instead of a period?

Author: Fixed

> . Creating symbolic links is faster than creating copies of the `/home/user` folder content in the persistent volume. To put it differently, the persistent volume in this case contains symbolic links and not the actual files themselves.

Reviewer 2: Though this might be a third advantage, you could even mention here that "Creating symbolic links is faster **and consumes less storage** than creating copies of the `/home/user` folder content in the persistent volume "

Author: Fixed

> . If the tools image is updated with newer versions of existing binaries, configs, and files, the init container does not need to `stow` the new versions, as the existing symbolic links will link to the newer versions in `/home/tooling` already.

Reviewer 2: Something that just occurred to me: if the tooling image is updated with **additional binaries, files etc.** they won't be copied to the `/home/user/` directory since the stow command won't be run again.

This may be something we want to document in the future (or somewhere in this document as a small note that the user should delete the` /home/user/.stow_completed` file and restart the workspace to pick up these additional files)

Author: I've added this note:
<img width="860" alt="image" src="https://github.com/user-attachments/assets/3ed73487-ff15-4f89-9d4f-60e996a4bb3f">

Reviewer 2: Perfect! Thank you :)

> 1. Creating symbolic links is faster than creating copies of the `/home/user` folder content in the persistent volume. To put it differently, the persistent volume in this case contains symbolic links and not the actual files themselves.
> 2. If the tools image is updated with newer versions of existing binaries, configs, and files, the init container does not need to `stow` the new versions, as the existing symbolic links will link to the newer versions in `/home/tooling` already.

Reviewer: ```suggestion
. Creating symbolic links is faster than creating copies of the `/home/user` folder content in the persistent volume. To put it differently, the persistent volume in this case contains symbolic links and not the actual files themselves.
. If the tools image is updated with newer versions of existing binaries, configs, and files, the init container does not need to `stow` the new versions, as the existing symbolic links will link to the newer versions in `/home/tooling` already.
```

> == Requirements for the tools image

Reviewer: ```suggestion
.`persistUserHome` tools image requirements
```

The `persistUserHome` depends on the tools image used for the workspace. By default {prod-short} uses the Universal Developer Image (UDI) for sample workspaces, which supports `persistUserHome` out of the box.

If you are using a custom image, there are three requirements that should be met to support the `persistUserHome` feature.

> . The tools image should contain `stow` version 2.4.0.

Reviewer 2: I'm not sure how often we'll update this document, so instead of setting a specific version of stow as a requirement, maybe we should say `version >= 2.4.0`?

Author: Fixed

> . The home directory should be `/home/user`.

Reviewer 2: Maybe this could be made more explicit?
"The home directory should be `/home/user`. **Meaning, the $HOME environment variable is set to `/home/user/`**"

Author: Fixed

> . In the tools image, the directory that contains the `/home/user` content is `/home/tooling`.

Reviewer 2: For clarity, maybe:

"In the tools image, the directory that **is intended to contain** the `/home/user` content is `/home/tooling`."?

Also, this sentence uses "directory" instead of "folder". I'd suggest sticking with one of the two terms for consistency in this doc.

Author: Fixed

> 1. The tools image should contain `stow` version 2.4.0.
> 2. The home directory should be `/home/user`.
> 3. In the tools image, the directory that contains the `/home/user` content is `/home/tooling`.

Reviewer: ```suggestion
. The tools image should contain `stow` version 2.4.0.
. The home directory should be `/home/user`.
. In the tools image, the directory that contains the `/home/user` content is `/home/tooling`.
```

Due to requirement three, the default UDI image for example adds the `/home/user` content to `/home/tooling` instead, and runs:

[subs="+quotes,attributes"]
----
RUN stow -t /home/user/ -d /home/tooling/ --no-folding 
----

> in the Dockerfile so that `/home/tooling` is accessible from `/home/user` even when not using the `persistUserHome` feature.

Reviewer 2: Maybe for clarity: "so that **files in** `/home/tooling` are accessible from `/home/user"?

Author: Fixed

RUN stow -t /home/user/ -d /home/tooling/ --no-folding 
----

> in the Dockerfile so that files in `/home/tooling` is accessible from `/home/user` even when not using the `persistUserHome` feature.

Reviewer 2: Sorry something I missed, "is" should be changed to "are" now:
"...so that files in `/home/tooling` *are* accessible from..."

Author: No problem, fixed
